### PR TITLE
Feature/seeded ksk

### PR DESCRIPTION
**Pull request by IceTDrinker**
### Resolves:

refs https://github.com/zama-ai/concrete-core-internal/issues/80
closes https://github.com/zama-ai/concrete-core-internal/issues/259

### Description

Seeded LweKeyswithKey

### Checklist 

(Use '[x]' to check the checkboxes, or submit the PR and then click the checkboxes)

* [x] Tests for the changes have been added (for bug fixes / features)
* [x] Docs have been added / updated (for bug fixes / features)
* [x] The PR description links to the related issue (to link an issue, use '#XXX'.)
* [x] The tests on AWS have been launched and are successful (apply the `aws_test` to the PR to launch the tests on AWS)
* [x] The draft release description has been updated
* [x] Check for breaking changes and add them to commit message following the conventional commit [specification][conventional-breaking]

[conventional-breaking]: https://www.conventionalcommits.org/en/v1.0.0/#commit-message-with-description-and-breaking-change-footer

### CODE REVIEW

**Comment by IceTDrinker** [comment]
@slab-ci cpu_test

**Comment by IceTDrinker** [comment]
updated with KSK in C FFI

**Comment by IceTDrinker** [comment]
rebased and added fixtures to bench (not tests as we don't yet have the keys statistical tests)

File: concrete-core/src/commons/crypto/lwe/seeded_keyswitch.rs [hunk 1]
```diff
@@ -0,0 +1,582 @@
...
+    ///     levels,
+    ///     base_log,
+    ///     input_dimension,
+    ///     output_dimension,
+    ///     compression_seed,
+    /// );
+    ///
+    /// assert_eq!(ksk.after_key_size(), output_dimension);
+    /// ```
+    pub fn after_key_size(&self) -> LweDimension
```
Comment (agnesLeroy) [inline concrete-core/src/commons/crypto/lwe/seeded_keyswitch.rs:128]: after_key_size -> output_lwe_dimension

File: concrete-core/src/commons/crypto/lwe/seeded_keyswitch.rs [hunk 2]
```diff
@@ -0,0 +1,582 @@
...
+    ///     levels,
+    ///     base_log,
+    ///     input_dimension,
+    ///     output_dimension,
+    ///     compression_seed,
+    /// );
+    ///
+    /// assert_eq!(ksk.decomposition_levels_count(), levels);
+    /// ```
+    pub fn decomposition_levels_count(&self) -> DecompositionLevelCount
```
Comment (agnesLeroy) [inline concrete-core/src/commons/crypto/lwe/seeded_keyswitch.rs:228]: I think we now use decomposition_level_count without plural for level.

File: concrete-core-fixture/src/fixture/lwe_seeded_keyswitch_to_lwe_keyswitch_key_transformation.rs [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+use crate::generation::{IntegerPrecision, Maker};
+use concrete_commons::dispersion::Variance;
+use concrete_commons::parameters::{DecompositionBaseLog, DecompositionLevelCount, LweDimension};
+use concrete_core::prelude::{
+    LweKeyswitchKeyEntity, LweSecretKeyEntity, LweSeededKeyswitchKeyEntity,
+    LweSeededKeyswitchToLweSeededKeyswitchKeyTransformationEngine,
+};
+
+/// A fixture for the types implementing the `LweSeededKeyswitchKeyCreationEngine` trait.
+pub struct LweSeededKeyswitchToLweKeyswitchKeyCreationFixture;
```
Comment (agnesLeroy) [inline concrete-core-fixture/src/fixture/lwe_seeded_keyswitch_to_lwe_keyswitch_key_transformation.rs:15]: Here it should be TransformationFixture, right?

File: concrete-core-fixture/src/fixture/lwe_seeded_keyswitch_to_lwe_keyswitch_key_transformation.rs [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+    Fixture<
+        Precision,
+        Engine,
+        (
+            InputSecretKey,
+            OutputSecretKey,
+            InputSeededKeyswitchKey,
+            OutputKeyswitchKey,
+        ),
+    > for LweSeededKeyswitchToLweKeyswitchKeyCreationFixture
```
Comment (agnesLeroy) [inline concrete-core-fixture/src/fixture/lwe_seeded_keyswitch_to_lwe_keyswitch_key_transformation.rs:43]: Same here

**Review by agnesLeroy — COMMENTED** [review]
Hi @IceTDrinker! Thanks a lot for this PR, very neat work as usual! :smiley: I just have some suggestions for naming in the private API (you followed the existing KSK API but it's actually deprecated so in my opinion it would be better to directly give names that fit with our new conventions).

File: concrete-core/src/commons/crypto/lwe/seeded_list.rs [hunk 5]
```diff
@@ -0,0 +1,279 @@
...
+    /// use concrete_core::commons::math::random::{CompressionSeed, Seed};
+    ///
+    /// let compression_seed = CompressionSeed { seed: Seed(42) };
+    ///
+    /// let list =
+    ///     LweSeededList::<Vec<u8>>::allocate(LweDimension(9), CiphertextCount(20), compression_seed);
+    ///
+    /// assert_eq!(list.get_compression_seed(), compression_seed);
+    /// ```
+    pub fn get_compression_seed(&self) -> CompressionSeed {
```
Comment (agnesLeroy) [inline concrete-core/src/commons/crypto/lwe/seeded_list.rs:177]: In the other getters we haven't included `get_` in the name, we should probably do the same here. Actually this holds for the LWE ciphertext & list as well sorry.

File: concrete-core/src/commons/crypto/lwe/seeded_keyswitch.rs [hunk 6]
```diff
@@ -0,0 +1,582 @@
...
+    compression_seed: CompressionSeed,
+}
+
+tensor_traits!(LweSeededKeyswitchKey);
+
+impl<Scalar> LweSeededKeyswitchKey<Vec<Scalar>>
+where
+    Scalar: Copy + Numeric,
+{
+    /// Allocates a seeded keyswitching key whose masks and bodies are all `value`.
```
[the remark this answers is not in the document]
Comment (IceTDrinker) [reply concrete-core/src/commons/crypto/lwe/seeded_keyswitch.rs:45]: will do !

File: concrete-core/src/commons/crypto/lwe/seeded_keyswitch.rs [hunk 7]
```diff
@@ -0,0 +1,582 @@
...
+            decomp_base_log,
+            decomp_level_count,
+            lwe_size: output_dimension.to_lwe_size(),
+            compression_seed,
+        }
+    }
+}
+
+impl<Cont> LweSeededKeyswitchKey<Cont> {
+    /// Return the size of the output key.
```
[the remark this answers is not in the document]
Comment (IceTDrinker) [reply concrete-core/src/commons/crypto/lwe/seeded_keyswitch.rs:102]: ok yeah, I don't remember if I used directly the original's PR as basis or re-used wording I found, I'll update

File: concrete-core-fixture/src/fixture/lwe_seeded_keyswitch_to_lwe_keyswitch_key_transformation.rs (hunk 3 above)
Comment (IceTDrinker) [reply concrete-core-fixture/src/fixture/lwe_seeded_keyswitch_to_lwe_keyswitch_key_transformation.rs:15]: you caught me ! I copy paste my fixtures before updating them 😅 , not sure how I managed to mix those !

File: concrete-core-fixture/src/fixture/lwe_seeded_keyswitch_to_lwe_keyswitch_key_transformation.rs [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+};
+use crate::generation::{IntegerPrecision, Maker};
+use concrete_commons::dispersion::Variance;
+use concrete_commons::parameters::{DecompositionBaseLog, DecompositionLevelCount, LweDimension};
+use concrete_core::prelude::{
+    LweKeyswitchKeyEntity, LweSecretKeyEntity, LweSeededKeyswitchKeyEntity,
+    LweSeededKeyswitchToLweSeededKeyswitchKeyTransformationEngine,
+};
+
+/// A fixture for the types implementing the `LweSeededKeyswitchKeyCreationEngine` trait.
```
[the remark this answers is not in the document]
Comment (IceTDrinker) [reply concrete-core-fixture/src/fixture/lwe_seeded_keyswitch_to_lwe_keyswitch_key_transformation.rs:14]: Ah damn, yes thanks !

File: concrete-core-ffi/src/backends/default/engines/destruction.rs [hunk 9]
```diff
@@ -299,6 +299,48 @@ pub unsafe extern "C" fn default_engine_destroy_lwe_keyswitch_key_unchecked_u64(
...
 
+/// Destroy an `LweSeededKeyswitchKey64`.
+///
+/// Refer to `concrete-core` implementation for detailed documentation.
+///
+/// This function is [checked](crate#safety-checked-and-unchecked-functions).
+#[no_mangle]
+pub unsafe extern "C" fn default_engine_destroy_lwe_seeded_keyswitch_key_u64(
+    engine: *mut DefaultEngine,
+    keyswitch_key: *mut LweSeededKeyswitchKey64,
```
[the remark this answers is not in the document]
Comment (IceTDrinker) [reply concrete-core-ffi/src/backends/default/engines/destruction.rs:310]: will update

File: concrete-core-ffi/tests/test_lwe_encrypt_keyswitch_w_ksk_serde_decrypt.c [hunk 10]
```diff
@@ -156,24 +177,43 @@ void keyswitch_unchecked_view_buffers_test(void) {
...
+      default_serialization_engine_serialize_lwe_seeded_keyswitch_key_unchecked_u64(
+          default_serialization_engine, seeded_ksk, &seeded_ksk_buffer);
+  assert(seeded_ksk_ser_ok == 0);
+
+  BufferView seeded_ksk_buffer_view = {.pointer = seeded_ksk_buffer.pointer,
+                                       .length = seeded_ksk_buffer.length};
+  LweSeededKeyswitchKey64 *deser_seeded_ksk = NULL;
+  int seeded_ksk_deser_ok =
+      default_serialization_engine_deserialize_lwe_seeded_keyswitch_key_unchecked_u64(
+          default_serialization_engine, seeded_ksk_buffer_view, &deser_seeded_ksk);
```
[the remark this answers is not in the document]
Comment (IceTDrinker) [reply concrete-core-ffi/tests/test_lwe_encrypt_keyswitch_w_ksk_serde_decrypt.c:197]: it's to convey the ownership notion in C IIRC

File: concrete-core/src/commons/crypto/lwe/seeded_list.rs (hunk 5 above)
Comment (IceTDrinker) [reply concrete-core/src/commons/crypto/lwe/seeded_list.rs:177]: not to worry 🙂

**Comment by IceTDrinker** [comment]
the last push takes into account your reviews @agnesLeroy I'm starting a second pass with the batch of comments from @aPere3

**Comment by IceTDrinker** [comment]
updated with the general comments from Alex.

This will still fail the clippy check until we merge #82

**Comment by IceTDrinker** [comment]
updated with a bit exact test of KSK and seeded KSK

**Comment by IceTDrinker** [comment]
rebased on main with the lwe ciphertext vector

**Pull request merged**